### PR TITLE
Feat/ai land extension

### PR DESCRIPTION
**Pull request by NRaoult**
Currently, the land component of the IFS (ecLand) is emulated using a simple MLP coded outside of Anemoi-core. 

To be able to utilise the full Anemoi workflow when emulating the land surface, we are adding a feature that would simplify the training in the anemoi-core, allowing for MLP-like training. This means effectively by-passing the graph structure by letting points be their own neighbours. This allows us to keep the single column structure of ecLand (no latent information between grids cells) and have a resolution agnostic model (i.e., if we train the model weights on an o96 dataset we can directly apply this to n320, as long as trainable_parameters is set to 0).

Next steps include by-passing the encoder/decoder structure by adding the option to have "0" num_heads.

### CODE REVIEW

File: training/pyproject.toml [hunk 1]
```diff
@@ -56,7 +56,7 @@ dependencies = [
   "pyshtools>=4.13",
   "pytorch-lightning>=2.1",
   "timm>=0.9.2",
-  "torch>=2.2",
+  "torch>=2.2,<=2.6",
```
Comment (mchantry) [inline training/pyproject.toml]: Is this related to the feature?

File: training/src/anemoi/training/losses/__init__.py [hunk 2]
```diff
@@ -26,5 +27,6 @@
     "MAELoss",
     "MSELoss",
     "RMSELoss",
+    "SmoothL1Loss",
```
Comment (mchantry) [inline training/src/anemoi/training/losses/__init__.py]: Ideally I think the loss should be brought in as a standalone PR.

File: models/src/anemoi/models/models/encoder_processor_decoder.py
Comment (mchantry) [inline models/src/anemoi/models/models/encoder_processor_decoder.py:1]: Why are there changes in this file?

File: models/src/anemoi/models/models/processor.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+from anemoi.models.distributed.shapes import get_shape_shards
+from anemoi.models.layers.graph import NamedNodesAttributes
+from anemoi.utils.config import DotDict
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AnemoiModelProc(nn.Module):
+    """Message passing graph neural network."""
```
Comment (mchantry) [inline models/src/anemoi/models/models/processor.py]: What type of models does this support? Purely local MLPs applied to each gripoint? Would be nice for the description to be expanded.

File: training/src/anemoi/training/config/debug_ailand.yaml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+- datamodule: single
+- diagnostics: evaluation
+- hardware: example
+- graph: ailand
+- model: graphtransformer
+- training: default
+- _self_
+
+
+config_validation: False
```
Comment (theissenhelen) [inline training/src/anemoi/training/config/debug_ailand.yaml]: My recommendation would be to set this to true, so that we can ensure that the default configs and schemas are functional for users who want to check their config.

File: training/src/anemoi/training/losses/__init__.py (hunk 2 above)
Comment (JPXKQX) [reply training/src/anemoi/training/losses/__init__.py]: Does something like https://github.com/ecmwf/anemoi-core/pull/398 make sense? @pinnstorm

File: models/src/anemoi/models/models/encoder_processor_decoder.py
Comment (JPXKQX) [reply models/src/anemoi/models/models/encoder_processor_decoder.py:1]: We should remove this, a separated PR was opened to delete these repeated lines

File: training/pyproject.toml (hunk 1 above)
Comment (JPXKQX) [reply training/pyproject.toml]: I don't think so

File: models/src/anemoi/models/models/processor.py (hunk 3 above)
Comment (JPXKQX) [reply models/src/anemoi/models/models/processor.py]: This class supports any of the previous processors (gnn, graphtransformer, transformer) and the new `PointMLP` at the data resolution, as it doesn't have an encoder/decoder.

File: training/src/anemoi/training/config/debug_ailand.yaml (hunk 4 above)
Comment (JPXKQX) [reply training/src/anemoi/training/config/debug_ailand.yaml]: Done!

**Comment by pinnstorm** [comment]
> Currently, the land component of the IFS (ecLand) is emulated using a simple MLP coded outside of Anemoi-core.
> 
> To be able to utilise the full Anemoi workflow when emulating the land surface, we are adding a feature that would simplify the training in the anemoi-core, allowing for MLP-like training. This means effectively by-passing the graph structure by letting points be their own neighbours. This allows us to keep the single column structure of ecLand (no latent information between grids cells) and have a resolution agnostic model (i.e., if we train the model weights on an o96 dataset we can directly apply this to n320, as long as trainable_parameters is set to 0).
> 
> Next steps include by-passing the encoder/decoder structure by adding the option to have "0" num_heads.

We can see some results from training a PointWiseMLP here:
https://mlflow.ecmwf.int/#/experiments/270/runs/c615f13596454d95ae6d7666fc54b90f/artifacts

This model specification will allow us to fine-tune a land surface model to point flux tower observations and potentially improve land-model fluxes will still being able to apply this model to a global grid

File: models/src/anemoi/models/models/processor.py (hunk 3 above)
Comment (ssmmnn11) [reply models/src/anemoi/models/models/processor.py]: Still says Message passing graph neural network

Comment (JPXKQX) [reply models/src/anemoi/models/models/processor.py]: Any suggestions? It can still be a messa gepassing gnn if you choose `processor: gnn`. What about a graph neural network?

Comment (ssmmnn11) [reply models/src/anemoi/models/models/processor.py]: I would recommend to use the encoder_processor_decoder structure with suitable encoder processor decoder instead.

File: models/tests/layers/mapper/test_graphinterpolation_mapper.py
Comment (mchantry) [inline models/tests/layers/mapper/test_graphinterpolation_mapper.py:1]: Empty file?

Comment (JPXKQX) [reply models/tests/layers/mapper/test_graphinterpolation_mapper.py:1]: It seems that this empty file was accidentally included in the main repository in a previous PR (#436).

File: training/docs/user-guide/models.rst [hunk 5]
```diff
@@ -99,6 +100,21 @@ coarser than the resolution of the base data.
...
+.. note::
+
+   The Transformer does not require a subgraph.
+
+Point-wise MLP
+==============
+
+The Point-wise MLP applies the same multilayer perceptron independently
+to each node. Because it ignores neighboring nodes (no message passing),
+the processor is location-independent.
```
Comment (ssmmnn11) [inline training/docs/user-guide/models.rst]: I think it would be better to state that is only point-wise, results are not conditioned on other notes. Location-independent is not well defined. If you provide positional information, it will learn from these.

File: models/src/anemoi/models/layers/block.py [hunk 6]
```diff
@@ -63,6 +63,35 @@ def forward(
...
+        super().__init__()
+        layers = [
+            layer_kernels.Linear(num_channels, hidden_dim),
+            layer_kernels.LayerNorm(hidden_dim),
+            layer_kernels.Activation(),
+        ]
+        if num_channels != hidden_dim:
+            layers.append(layer_kernels.Linear(hidden_dim, num_channels))
+
+        if dropout_p is not None and dropout_p > 0:
```
Comment (theissenhelen) [inline models/src/anemoi/models/layers/block.py:81]: It might better to raise an error for `dropout <0` instead of letting it fall through.

File: models/src/anemoi/models/layers/block.py [hunk 7]
```diff
@@ -63,6 +63,35 @@ def forward(
...
 
 
+class PointWiseMLPProcessorBlock(BaseBlock):
+    """Point-wise block with MLPs."""
+
+    def __init__(self, *, num_channels: int, hidden_dim: int, layer_kernels: DotDict, dropout_p: float = 0.0):
+        super().__init__()
+        layers = [
+            layer_kernels.Linear(num_channels, hidden_dim),
+            layer_kernels.LayerNorm(hidden_dim),
```
Comment (ssmmnn11) [inline models/src/anemoi/models/layers/block.py:75]: Why do we layernorm after linear and before activation? Is this a intended pattern? :-)

File: models/src/anemoi/models/layers/block.py (hunk 6 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/block.py:81]: Thanks, check added now! I think the purpose of setting `dropout_p > 0` here was to avoid instantiating a Dropout(0).

Comment (ssmmnn11) [inline models/src/anemoi/models/layers/block.py:81]: please note that dropout is currently not safe to use for sharded models.

File: training/src/anemoi/training/config/model/point_wise.yaml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    bias: False
+
+processor:
+  _target_: anemoi.models.layers.processor.PointWiseMLPProcessor
+  trainable_size: ${model.trainable_parameters.hidden2hidden}
+  sub_graph_edge_attributes: ${model.attributes.edges}
+  num_layers: 16
+  num_chunks: 2
+  mlp_hidden_ratio: 4 # GraphTransformer or Transformer only
+  num_heads: 16 # GraphTransformer or Transformer only
```
Comment (ssmmnn11) [inline training/src/anemoi/training/config/model/point_wise.yaml]: I don't hink you have heads here

File: training/src/anemoi/training/config/model/point_wise.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  qk_norm: False
```
Comment (ssmmnn11) [inline training/src/anemoi/training/config/model/point_wise.yaml]: qk norm is only relevant for the transformer graphtransformer

File: training/src/anemoi/training/config/model/point_wise.yaml (hunk 8 above)
Comment (JPXKQX) [reply training/src/anemoi/training/config/model/point_wise.yaml]: You are right! It is removed now

File: models/src/anemoi/models/layers/block.py (hunk 7 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/block.py:75]: Do you have any comments on this? @pinnstorm

File: training/docs/user-guide/models.rst (hunk 5 above)
Comment (JPXKQX) [reply training/docs/user-guide/models.rst]: The documentation has now been updated

File: models/src/anemoi/models/layers/block.py (hunk 7 above)
Comment (pinnstorm) [reply models/src/anemoi/models/layers/block.py:75]: This was just one of the patterns we had tested with the simple external MLP that had given us good results for land-surface model emulation. But don't think the exploration was exhaustive so if we want something that conforms more with the other MLPs we have here sure that is fine too

File: models/src/anemoi/models/layers/block.py (hunk 6 above)
Comment (JPXKQX) [reply models/src/anemoi/models/layers/block.py:81]: What would you recommend here?

Comment (anaprietonem) [reply models/src/anemoi/models/layers/block.py:81]: I have added now an assert for this and included a unit test to check that if sharding and dropout are used an error is raised. Thoughts?

File: models/src/anemoi/models/schemas/processor.py [hunk 10]
```diff
@@ -79,3 +79,14 @@ def check_valid_extras(self) -> Any:
...
+
+class PointWiseMLPProcessorSchema(PointWiseModelComponent):
+    target_: Literal["anemoi.models.layers.processor.PointWiseMLPProcessor"] = Field(..., alias="_target_")
+    "Transformer processor object from anemoi.models.layers.processor."
+    num_layers: NonNegativeInt = Field(example=16)
+    "Number of layers of Transformer processor."
+    num_channels: NonNegativeInt = Field(example=128)
+    "Number of channels."
+    dropout_p: NonNegativeFloat = Field(example=0.1)
+    "Dropout probability used for multi-head self attention, default 0.0"
```
Comment (ssmmnn11) [inline models/src/anemoi/models/schemas/processor.py]: Remove multi-head self attention?

Comment (JPXKQX) [reply models/src/anemoi/models/schemas/processor.py]: Removed!

**Review by ssmmnn11 — APPROVED** [review]
I think looks good.
